### ISSUE TITLE
History - Allow "First N" query

### ISSUE THREAD

**Issue by bizetti** [dojot/dojot#1701]
It returns the first N registers in the data base (in a time interval, if it is mentioned).

"First N" means the N oldest registers observing their timestamps.

**Comment by thiagodpf** [comment]
link to PRs:
https://github.com/dojot/history/pull/91
https://github.com/dojot/history/pull/92

**Comment by betefaber** [comment]
Fixed in **v0.4.4-rc.1**.

-all values:
```
curl -X GET 10.202.71.7:8000/history/device/d21292/history -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
{"temperatura": [
{"attr": "temperatura", "value": 20, "device_id": "d21292", "ts": "2020-10-19T13:46:42.327000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 10, "device_id": "d21292", "ts": "2020-10-19T13:46:36.716000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 30, "device_id": "d21292", "ts": "2020-10-19T13:46:26.280000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 40, "device_id": "d21292", "ts": "2020-10-19T13:46:10.920000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 40, "device_id": "d21292", "ts": "2020-10-19T13:46:01.992000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 50, "device_id": "d21292", "ts": "2020-10-19T13:45:09.025000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 20, "device_id": "d21292", "ts": "2020-10-19T13:44:44.741000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 20, "device_id": "d21292", "ts": "2020-10-19T13:44:10.014000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 10, "device_id": "d21292", "ts": "2020-10-19T13:43:20.540000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 30, "device_id": "d21292", "ts": "2020-10-19T13:27:21.854000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 10, "device_id": "d21292", "ts": "2020-10-19T13:25:58.406000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 20, "device_id": "d21292", "ts": "2020-10-19T12:53:57.332000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 20, "device_id": "d21292", "ts": "2020-10-19T12:52:07.059000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 10, "device_id": "d21292", "ts": "2020-10-19T12:51:01.731000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 30, "device_id": "d21292", "ts": "2020-10-19T12:39:20.276000Z", "metadata": {}}
], 
"protocol": []}
```
- filter by last N values
```
curl -X GET 10.202.71.7:8000/history/device/d21292/history?lastN=3 -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
{"temperatura": [
{"attr": "temperatura", "value": 20, "device_id": "d21292", "ts": "2020-10-19T13:46:42.327000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 10, "device_id": "d21292", "ts": "2020-10-19T13:46:36.716000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 30, "device_id": "d21292", "ts": "2020-10-19T13:46:26.280000Z", "metadata": {}}
], 
"protocol": []}
```
- filter by **first N** values
```
curl -X GET 10.202.71.7:8000/history/device/d21292/history?firstN=3 -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
{"temperatura": [
{"attr": "temperatura", "value": 30, "device_id": "d21292", "ts": "2020-10-19T12:39:20.276000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 10, "device_id": "d21292", "ts": "2020-10-19T12:51:01.731000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 20, "device_id": "d21292", "ts": "2020-10-19T12:52:07.059000Z", "metadata": {}}
], 
"protocol": []}
```
- filter by all parameters
```
curl -X GET '10.202.71.7:8000/history/device/d21292/history?firstN=3&attr=temperatura&dateFrom=2020-10-01T00:00:00Z&dateTo=2020-10-19T14:00:00Z' -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
[
{"attr": "temperatura", "value": 30, "device_id": "d21292", "ts": "2020-10-19T12:39:20.276000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 10, "device_id": "d21292", "ts": "2020-10-19T12:51:01.731000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 20, "device_id": "d21292", "ts": "2020-10-19T12:52:07.059000Z", "metadata": {}}
]
```
```
curl -X GET '10.202.71.7:8000/history/device/d21292/history?lastN=3&attr=temperatura&dateFrom=2020-10-01T00:00:00Z&dateTo=2020-10-19T14:00:00Z' -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
[
{"attr": "temperatura", "value": 20, "device_id": "d21292", "ts": "2020-10-19T13:46:42.327000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 10, "device_id": "d21292", "ts": "2020-10-19T13:46:36.716000Z", "metadata": {}}, 
{"attr": "temperatura", "value": 30, "device_id": "d21292", "ts": "2020-10-19T13:46:26.280000Z", "metadata": {}}
]
```
- no history in these periods
```
curl -X GET '10.202.71.7:8000/history/device/d21292/history?firstN=3&dateFrom=2020-10-19T00:00:00Z&dateTo=2020-10-19T10:00:00Z' -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
{"temperatura": [], "protocol": []}
```
```
 curl -X GET '10.202.71.7:8000/history/device/d21292/history?firstN=3&dateFrom=2020-10-01T00:00:00Z&dateTo=2020-10-14T14:00:00Z' -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
{"temperatura": [], "protocol": []}
```
- no history for that attribute yet
```
curl -X GET '10.202.71.7:8000/history/device/d21292/history?firstN=3&attr=temperatura&dateFrom=2020-10-01T00:00:00Z&dateTo=2020-10-14T14:00:00Z' -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
{
    "title": "Attr not found",
    "description": "No data for the given attribute could be found"
}
```

**Comment by betefaber** [comment]
Tested in **v0.5.0-rc.1**

- all values

```
curl -X GET "${DOJOT_HOST}/history/device/6be244/history?attr=float" -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
[
{"attr": "float", "value": 90000, "device_id": "6be244", "ts": "2020-11-09T15:11:14.154000Z", "metadata": {}}, 
{"attr": "float", "value": 80000, "device_id": "6be244", "ts": "2020-11-09T15:11:08.994000Z", "metadata": {}}, 
{"attr": "float", "value": 70000, "device_id": "6be244", "ts": "2020-11-09T15:11:04.394000Z", "metadata": {}}, 
{"attr": "float", "value": 60000, "device_id": "6be244", "ts": "2020-11-09T15:11:00.665000Z", "metadata": {}}, 
{"attr": "float", "value": 50000, "device_id": "6be244", "ts": "2020-11-09T15:10:52.120000Z", "metadata": {}}, 
{"attr": "float", "value": 40000, "device_id": "6be244", "ts": "2020-11-09T15:10:47.420000Z", "metadata": {}}, 
{"attr": "float", "value": 30000, "device_id": "6be244", "ts": "2020-11-09T15:10:43.634000Z", "metadata": {}}, 
{"attr": "float", "value": 20000, "device_id": "6be244", "ts": "2020-11-09T15:10:40.067000Z", "metadata": {}}, 
{"attr": "float", "value": 10000, "device_id": "6be244", "ts": "2020-11-09T15:10:35.839000Z", "metadata": {}}, 
{"attr": "float", "value": 9000, "device_id": "6be244", "ts": "2020-11-09T15:04:36.258000Z", "metadata": {}}, 
{"attr": "float", "value": 8000, "device_id": "6be244", "ts": "2020-11-09T15:04:29.374000Z", "metadata": {}}, 
{"attr": "float", "value": 7000, "device_id": "6be244", "ts": "2020-11-09T15:04:10.173000Z", "metadata": {}}, 
{"attr": "float", "value": 6000, "device_id": "6be244", "ts": "2020-11-09T15:04:05.281000Z", "metadata": {}}, 
{"attr": "float", "value": 5000, "device_id": "6be244", "ts": "2020-11-09T15:04:01.521000Z", "metadata": {}}, 
{"attr": "float", "value": 4000, "device_id": "6be244", "ts": "2020-11-09T15:03:57.769000Z", "metadata": {}}, 
{"attr": "float", "value": 3000, "device_id": "6be244", "ts": "2020-11-09T15:03:54.539000Z", "metadata": {}}, 
{"attr": "float", "value": 2000, "device_id": "6be244", "ts": "2020-11-09T15:03:49.675000Z", "metadata": {}}, 
{"attr": "float", "value": 1000, "device_id": "6be244", "ts": "2020-11-09T15:03:45.377000Z", "metadata": {}}
]
```
- filter by last N values
```
curl -X GET "${DOJOT_HOST}/history/device/6be244/history?lastN=3&attr=float" -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
[
{"attr": "float", "value": 90000, "device_id": "6be244", "ts": "2020-11-09T15:11:14.154000Z", "metadata": {}}, 
{"attr": "float", "value": 80000, "device_id": "6be244", "ts": "2020-11-09T15:11:08.994000Z", "metadata": {}}, 
{"attr": "float", "value": 70000, "device_id": "6be244", "ts": "2020-11-09T15:11:04.394000Z", "metadata": {}}
]
```
- filter by first N values
```
curl -X GET "${DOJOT_HOST}/history/device/6be244/history?firstN=3&attr=float" -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
[
{"attr": "float", "value": 1000, "device_id": "6be244", "ts": "2020-11-09T15:03:45.377000Z", "metadata": {}}, 
{"attr": "float", "value": 2000, "device_id": "6be244", "ts": "2020-11-09T15:03:49.675000Z", "metadata": {}}, 
{"attr": "float", "value": 3000, "device_id": "6be244", "ts": "2020-11-09T15:03:54.539000Z", "metadata": {}}
]
```
